File: flutter/lib/data/repositories/url_repository_impl.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';
import '../../domain/entities/cleaned_url.dart';
import '../../domain/entities/url_parameter.dart';
import '../../domain/repositories/url_repository.dart';

class UrlRepositoryImpl implements UrlRepository {
  static const String _historyKey = 'url_cleaning_history';
  static const String _statsKey = 'url_cleaning_stats';
  static const int _maxHistoryItems = 100;

  final SharedPreferences _prefs;

  UrlRepositoryImpl(this._prefs);

  @override
  Future<CleanedUrl> parseUrl(String url) async {
    final uri = Uri.parse(url);
    final parameters = uri.queryParameters.entries.map((entry) {
      return UrlParameter(
        key: entry.key,
        value: entry.value,
        selected: true,
        isTracker: _isCommonTracker(entry.key),
        category: _getTrackerCategory(entry.key),
      );
    }).toList();

    return CleanedUrl.fromUrl(url, parameters);
  }

  @override
  Future<void> saveToHistory(CleanedUrl cleanedUrl) async {
    final history = await getHistory();
    history.insert(0, cleanedUrl);
    
    if (history.length > _maxHistoryItems) {
      history.removeRange(_maxHistoryItems, history.length);
    }

    final jsonList = history.map((item) => _cleanedUrlToJson(item)).toList();
    await _prefs.setStringList(_historyKey, jsonList.map((e) => jsonEncode(e)).toList());
    
    await _updateStatistics(cleanedUrl);
  }

  @override
  Future<List<CleanedUrl>> getHistory() async {
    final jsonList = _prefs.getStringList(_historyKey) ?? [];
    return jsonList
        .map((json) => _cleanedUrlFromJson(jsonDecode(json)))
        .toList();
  }

  @override
  Future<void> clearHistory() async {
    await _prefs.remove(_historyKey);
    await _prefs.remove(_statsKey);
  }

  @override
  Future<Map<String, int>> getStatistics() async {
    final statsJson = _prefs.getString(_statsKey);
    if (statsJson == null) {
      return {
        'totalUrls': 0,
        'totalCharactersSaved': 0,
        'totalTrackersRemoved': 0,
      };
    }
    return Map<String, int>.from(jsonDecode(statsJson));
  }

  Future<void> _updateStatistics(CleanedUrl cleanedUrl) async {
    final stats = await getStatistics();
    stats['totalUrls'] = (stats['totalUrls'] ?? 0) + 1;
    stats['totalCharactersSaved'] = (stats['totalCharactersSaved'] ?? 0) + cleanedUrl.charactersSaved;
    stats['totalTrackersRemoved'] = (stats['totalTrackersRemoved'] ?? 0) + 
        cleanedUrl.parameters.where((p) => p.isTracker && !p.selected).length;
    
    await _prefs.setString(_statsKey, jsonEncode(stats));
  }

  Map<String, dynamic> _cleanedUrlToJson(CleanedUrl url) {
    return {
      'originalUrl': url.originalUrl,
      'baseUrl': url.baseUrl,
      'cleanedUrl': url.cleanedUrl,
      'parameters': url.parameters.map((p) => {
        'key': p.key,
        'value': p.value,
        'selected': p.selected,
        'isTracker': p.isTracker,
        'category': p.category,
      }).toList(),
      'charactersSaved': url.charactersSaved,
      'cleanedAt': url.cleanedAt.toIso8601String(),
    };
  }

  CleanedUrl _cleanedUrlFromJson(Map<String, dynamic> json) {
    final parameters = (json['parameters'] as List).map((p) {
      return UrlParameter(
        key: p['key'],
        value: p['value'],
        selected: p['selected'],
        isTracker: p['isTracker'],
        category: p['category'],
      );
    }).toList();

    return CleanedUrl(
      originalUrl: json['originalUrl'],
      baseUrl: json['baseUrl'],
      cleanedUrl: json['cleanedUrl'],
      parameters: parameters,
      charactersSaved: json['charactersSaved'],
      cleanedAt: DateTime.parse(json['cleanedAt']),
    );
  }

  bool _isCommonTracker(String key) {
    const trackers = {
      'utm_source', 'utm_medium', 'utm_campaign', 'utm_term', 'utm_content',
      'fbclid', 'gclid', 'msclkid', 'ref', 'referrer', 'source',
      'campaign', 'ad', 'adgroup', 'creative', 'keyword',
      '_ga', '_gid', '__hstc', '__hssc', '__hsfp',
      'mc_cid', 'mc_eid', 'mkt_tok', 'trk', 'trkInfo',
    };
    return trackers.contains(key.toLowerCase());
  }

  String? _getTrackerCategory(String key) {
    const categories = {
      'utm_source': 'Google Analytics',
      'utm_medium': 'Google Analytics',
      'utm_campaign': 'Google Analytics',
      'fbclid': 'Facebook',
      'gclid': 'Google Ads',
      'msclkid': 'Microsoft',
      '_ga': 'Google Analytics',
      'mc_cid': 'Mailchimp',
      '__hstc': 'HubSpot',
    };
    return categories[key.toLowerCase()];
  }
}

File: react_native/src/data/repositories/UrlRepositoryImpl.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { UrlRepository, UrlStatistics } from '../../domain/repositories/UrlRepository';
import { CleanedUrl, createCleanedUrl } from '../../domain/entities/CleanedUrl';
import { UrlParameter } from '../../domain/entities/UrlParameter';
import { COMMON_TRACKERS } from '../../utils/constants';
import { ParameterIdentificationService } from '../../services/ParameterIdentificationService';

const HISTORY_KEY = '@link_cleaner_history';
const STATS_KEY = '@link_cleaner_stats';
const MAX_HISTORY_ITEMS = 100;

export class UrlRepositoryImpl implements UrlRepository {
  async parseUrl(url: string): Promise<CleanedUrl> {
    const urlObj = new URL(url);
    const parameters: UrlParameter[] = [];
    
    urlObj.searchParams.forEach((value, key) => {
      const paramInfo = ParameterIdentificationService.identifyParameter(key, url);
      
      parameters.push({
        key,
        value,
        selected: true,
        isTracker: paramInfo.isTracker,
        category: paramInfo.category,
        displayName: paramInfo.name,
        description: paramInfo.description,
      });
    });
    
    return createCleanedUrl(url, parameters);
  }

  async saveToHistory(cleanedUrl: CleanedUrl): Promise<void> {
    const history = await this.getHistory();
    history.unshift(cleanedUrl);
    
    if (history.length > MAX_HISTORY_ITEMS) {
      history.splice(MAX_HISTORY_ITEMS);
    }
    
    await AsyncStorage.setItem(HISTORY_KEY, JSON.stringify(history));
    await this.updateStatistics(cleanedUrl);
  }

  async getHistory(): Promise<CleanedUrl[]> {
    try {
      const historyJson = await AsyncStorage.getItem(HISTORY_KEY);
      if (!historyJson) return [];
      
      const history = JSON.parse(historyJson);
      return history.map((item: any) => ({
        ...item,
        cleanedAt: new Date(item.cleanedAt),
      }));
    } catch (error) {
      console.error('Error loading history:', error);
      return [];
    }
  }

  async clearHistory(): Promise<void> {
    await AsyncStorage.multiRemove([HISTORY_KEY, STATS_KEY]);
  }

  async getStatistics(): Promise<UrlStatistics> {
    try {
      const statsJson = await AsyncStorage.getItem(STATS_KEY);
      if (!statsJson) {
        return {
          totalUrls: 0,
          totalCharactersSaved: 0,
          totalTrackersRemoved: 0,
        };
      }
      return JSON.parse(statsJson);
    } catch (error) {
      console.error('Error loading statistics:', error);
      return {
        totalUrls: 0,
        totalCharactersSaved: 0,
        totalTrackersRemoved: 0,
      };
    }
  }

  private async updateStatistics(cleanedUrl: CleanedUrl): Promise<void> {
    const stats = await this.getStatistics();
    
    stats.totalUrls++;
    stats.totalCharactersSaved += cleanedUrl.charactersSaved;
    stats.totalTrackersRemoved += cleanedUrl.parameters.filter(
      p => p.isTracker && !p.selected
    ).length;
    
    await AsyncStorage.setItem(STATS_KEY, JSON.stringify(stats));
  }

  private getTrackerCategory(key: string): string | undefined {
    const categories: Record<string, string> = {
      'utm_source': 'Google Analytics',
      'utm_medium': 'Google Analytics',
      'utm_campaign': 'Google Analytics',
      'utm_term': 'Google Analytics',
      'utm_content': 'Google Analytics',
      'fbclid': 'Facebook',
      'gclid': 'Google Ads',
      'msclkid': 'Microsoft',
      'twclid': 'Twitter',
      'ttclid': 'TikTok',
      'mc_cid': 'Mailchimp',
      'mc_eid': 'Mailchimp',
      '__hstc': 'HubSpot',
      '__hssc': 'HubSpot',
      '__hsfp': 'HubSpot',
      '_ga': 'Google Analytics',
      '_gid': 'Google Analytics',
    };
    
    return categories[key.toLowerCase()];
  }
}